client/task_runner.go

```diff
@@ -166,94 +179,115 @@ func (r *TaskRunner) startTask() error {
 	if err != nil {
 		r.logger.Printf("[ERR] client: failed to start task '%s' for alloc '%s': %v",
 			r.task.Name, r.allocID, err)
-		r.setStatus(structs.AllocClientStatusFailed,
-			fmt.Sprintf("failed to start: %v", err))
+		e := structs.NewTaskEvent(structs.TaskDriverFailure).
+			SetDriverError(fmt.Errorf("failed to start: %v", err))
+		r.setState(structs.TaskStateDead, e)
 		return err
 	}
 	r.handle = handle
-	r.setStatus(structs.AllocClientStatusRunning, "task started")
+	r.setState(structs.TaskStateRunning, structs.NewTaskEvent(structs.TaskStarted))
 	return nil
 }
 
 // Run is a long running routine used to manage the task
 func (r *TaskRunner) Run() {
-	var err error
 	defer close(r.waitCh)
 	r.logger.Printf("[DEBUG] client: starting task context for '%s' (alloc '%s')",
 		r.task.Name, r.allocID)
 
-	// Start the task if not yet started
-	if r.handle == nil {
-		if err := r.startTask(); err != nil {
-			return
-		}
-	}
+	r.run(false)
```

Comment:
Will the stack ever become very deep if we restart in a hot loop when the user restarts with a very little delay and restart interval and the task segfaults?